library/ssl_tls13_client.c

```diff
@@ -401,25 +401,24 @@ static int ssl_tls13_read_public_ecdhe_share( mbedtls_ssl_context *ssl,
                                               size_t buf_len )
 {
     int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    uint8_t ecpoint_len;
+    uint8_t *p = (uint8_t*)buf;
+    mbedtls_ssl_handshake_params *handshake = ssl->handshake;
 
-    ret = mbedtls_ecdh_tls13_read_public( &ssl->handshake->ecdh_ctx,
-                                          buf, buf_len );
-    if( ret != 0 )
-    {
-        MBEDTLS_SSL_DEBUG_RET( 1, ( "mbedtls_ecdh_tls13_read_public" ), ret );
+    /*
+     * Put peer's ECDH public key in the format understood by PSA.
+     */
 
-        MBEDTLS_SSL_PEND_FATAL_ALERT( MBEDTLS_SSL_ALERT_MSG_ILLEGAL_PARAMETER,
-                                      MBEDTLS_ERR_SSL_ILLEGAL_PARAMETER );
-        return( MBEDTLS_ERR_SSL_ILLEGAL_PARAMETER );
-    }
+    ecpoint_len = *p++;
+    if( ( buf_len - 1 ) < ecpoint_len )
+        return( MBEDTLS_ERR_SSL_DECODE_ERROR );
 
-    if( ssl_tls13_check_ecdh_params( ssl ) != 0 )
+    if ( ( ret = mbedtls_psa_tls_ecpoint_to_psa_ec( p,
```

Comment:
This seems to be wrong way of reading key from server.

Reply:
According to [tls 1.3 spec](https://datatracker.ietf.org/doc/html/rfc8446#section-4.2.8) key_shares extension is built as follows:
```
      struct {
          NamedGroup group;
          opaque key_exchange<1..2^16-1>;
      } KeyShareEntry;
```

`group` is already processed, so here buf points to `opaque key_exchange<1..2^16-1>;`. I can't find how to import opaque key into the `handshake->ecdh_psa_peerkey` so it can be then used for `psa_raw_key_agreement()`.

While debugging I noticed that expected key size is 32 bytes and `buf` consists of 34 bytes where first two bytes are `0x00 0x20(32d)` so it looks like size and then we would have 32 bytes of data.

Reply:
The wire representation of `opaque key_exchange<1..2^16-1>` is indeed a two-byte size followed by the data, per [§3.4](https://datatracker.ietf.org/doc/html/rfc8446#section-3.4).

The contents of the `key_exchange` value depend on the group. It's 32 bytes for X25519 ([near the end of §4.2.8.2](https://datatracker.ietf.org/doc/html/rfc8446#section-4.2.8.2)).

Reply:
Note: the function `mbedtls_psa_tls_ecpoint_to_psa_ec()` is removed in https://github.com/ARMmbed/mbedtls/pull/5456. The description of #5321 clearly states that it depends on #5291, solved by #5456. Please base your work on #5456 (and label this PR "needs: preceding PR).

Reply:
In current version `mbedtls_psa_tls_ecpoint_to_psa_ec()` is not used. 
BTW can we merge https://github.com/ARMmbed/mbedtls/pull/5456 ? It is approved and CI is green.

Reply:
As the labels indicate, 5456 is waiting for the 2.28 backport (of the bug fix) (which I just created this morning). We usually wait for backports to be ready before merging a PR, because otherwise the risk of forgetting about the backport is just too big.

Reply:
I reviewed the backport and no issues found. It is now accepted by two reviewers, so it seems we are good to go with both PRs.